Packs/PAN-OS/Integrations/Panorama/Panorama.py

```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+
+        return topology
+
+    def freeze(self) -> dict:
+        """
+        Serializes the topology information for storage (for example, in the incident context)
+        """
+        frozen_topology = FrozenTopology
+        panorama_object: Panorama
+        for panorama_object in self.panorama_objects:
+            frozen_topology.panorama_objects.append(panorama_object.hostname)
+        firewall_object: Firewall
+        for firewall_object in self.panorama_objects:
+            frozen_topology.firewall_objects.append(firewall_object.hostname)
+
+        return vars(frozen_topology)
+
+    def get_direct_device(self, firewall: Firewall) -> PanDevice:
+        """
+        Given a firewall object that's proxied via Panorama, create a device that uses a direct API connection
+        instead. Used by any command that can't be routed via Panorama.
+        """
+        if firewall.hostname:
+            # If it's already a direct connection
+            return firewall
+
+        ip_address = firewall.show_system_info().get("system").get("ip-address")
+
+        return PanDevice.create_from_device(
+            hostname=ip_address,
+            api_username=self.username,
+            api_password=self.password
+        )
+
+    def get_all_object_containers(
+            self,
+            device_filter_string: str = None,
+            container_name: str = None,
+            top_level_devices_only: bool = False,
+    ) -> \
+            List[tuple[PanDevice, Union[Panorama, Firewall, DeviceGroup, Template, Vsys]]]:
+        """
+        Given a device, returns all the possible configuration containers that can contain objects -
+        vsys, device-groups, templates and template-stacks.
+        """
+        containers: List[Union[tuple[PanDevice, Callable], tuple[PanDevice, PanDevice]]] = []
+        # for device in self.all(device_filter_string):
+        # Changed to only refer to active devices, no passives.
+        device_retrieval_func = self.active_devices
+        if top_level_devices_only:
+            device_retrieval_func = self.active_top_level_devices  # type: ignore
+
+        for device in device_retrieval_func(device_filter_string):
+            device_groups = DeviceGroup.refreshall(device)
+            for device_group in device_groups:
+                containers.append((device, device_group))
+
+            templates = Template.refreshall(device)
+            for template in templates:
+                containers.append((device, template))
+
+            virtual_systems = Vsys.refreshall(device)
+            for virtual_system in virtual_systems:
+                containers.append((device, virtual_system))
+
+            if isinstance(device, Panorama):
+                # Add the "shared" device if Panorama. Firewalls will always have vsys1
+                containers.append((device, device))
+
+        return_containers = []
+
+        if container_name:
+            for container in containers:
```

Comment:
the ```containers``` is a tuple of two values?

Reply:
containers is a list of tuple. The list itself can be variable length